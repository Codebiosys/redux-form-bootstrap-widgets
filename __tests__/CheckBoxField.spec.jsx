/* eslint-disable import/first */

import React from 'react';
import { mount } from 'enzyme';
import { omit } from 'lodash';

import { CheckBoxField } from 'index';

const fieldProps = {
  input: {
    name: 'fieldName',
    onChange: jest.fn(),
    onBlur: jest.fn(),
    onFocus: jest.fn(),
  },
  meta: {
    form: 'formName',
    pristine: false,
  },
  label: 'Checkbox Test',
  helpText: 'The help Text',
  options: [
    { label: 'Option 1', value: 'One' },
    { label: 'Option 2', value: 'Two' },
    { label: 'Option 3', value: 'Three' },
  ],
};

const selectedFieldProps = { ...fieldProps };
selectedFieldProps.input = { ...fieldProps.input };
selectedFieldProps.input.value = ['One'];

const customOptionFieldProps = { ...fieldProps };
customOptionFieldProps.labelKey = 'foo';
customOptionFieldProps.valueKey = 'bar';
customOptionFieldProps.options = [
  { foo: 'Option 1', bar: 'One' },
  { foo: 'Option 2', bar: 'Two' },
  { foo: 'Option 3', bar: 'Three' }];

describe('The Checkbox Field', () => {
  let inputWrapper;

  beforeEach(() => {
    inputWrapper = mount(<CheckBoxField {...fieldProps} />);
  });

  it('renders', () => {
    expect(inputWrapper).toMatchSnapshot();
  });

  it('has a label when there is a label', () => {
    expect(inputWrapper.find('ControlLabel').text()).toEqual(fieldProps.label);
  });

  it('does not have a label when there is no label', () => {
    const noLabelFieldProps = omit(fieldProps, ['label']);
    const inputWrapperNoLabel = mount(<CheckBoxField {...noLabelFieldProps} />);
    expect(inputWrapperNoLabel.find('Label').exists()).toBe(false);
  });

  it('renders with custom options', () => {
    inputWrapper = mount(<CheckBoxField {...customOptionFieldProps} />);
    expect(inputWrapper).toMatchSnapshot();
  });
  it('triggers the focus when focused', () => {
    inputWrapper.find(`input[name="${fieldProps.input.name}_0"]`).simulate('focus');
    expect(fieldProps.input.onFocus).toHaveBeenCalledTimes(1);
  });

  it('calls onChange and onBlur when the checkbox is selected', () => {
    inputWrapper.find(`input[name="${fieldProps.input.name}_0"]`).simulate('change', { target: { checked: true } });
    expect(fieldProps.input.onChange).toHaveBeenCalledTimes(1);
    expect(fieldProps.input.onBlur).toHaveBeenCalledTimes(1);
    expect(fieldProps.input.onChange).toHaveBeenCalledWith(['One']);
  });

  it('calls onChange and onBlur when the checkbox is deselected', () => {
    inputWrapper = mount(<CheckBoxField {...selectedFieldProps} />);

    inputWrapper.find(`input[name="${fieldProps.input.name}_0"]`).simulate('change', { target: { checked: false } });
    expect(fieldProps.input.onChange).toHaveBeenCalledTimes(1);
    expect(fieldProps.input.onBlur).toHaveBeenCalledTimes(1);
    expect(fieldProps.input.onChange).toHaveBeenCalledWith(null);
  });

  it('calls custom validator when the toogle is toggled', () => {
    const customValidator = jest.fn(() => ({ validationState: null, errorMessage: null }));
    const validatorProps = { ...fieldProps, customValidation: customValidator };
    const inputWrapperValidated = mount(<CheckBoxField {...validatorProps} />);

    inputWrapperValidated.find(`input[name="${fieldProps.input.name}_0"]`).simulate('change', { target: { checked: false } });
    expect(customValidator).toHaveBeenCalledTimes(1);
  });
});
