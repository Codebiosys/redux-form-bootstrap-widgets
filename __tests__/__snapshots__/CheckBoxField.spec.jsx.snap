// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Checkbox Field renders 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="fieldName"
>
  <Label
    label="Checkbox Test"
    required={false}
  />
  <InputGroup
    bsClass="input-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      inline={false}
      name="fieldName_option1"
      onChange={[Function]}
      onFocus={[Function]}
      title=""
      value="One"
    >
      Option 1
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      inline={false}
      name="fieldName_option2"
      onChange={[Function]}
      onFocus={[Function]}
      title=""
      value="Two"
    >
      Option 2
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      inline={false}
      name="fieldName_option3"
      onChange={[Function]}
      onFocus={[Function]}
      title=""
      value="Three"
    >
      Option 3
    </Checkbox>
  </InputGroup>
  <HelpBlock
    bsClass="help-block"
    style={
      Object {
        "minHeight": "3ex",
      }
    }
  />
</FormGroup>
`;

exports[`The Checkbox Field renders the help message with a break 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="fieldName"
  validationState={null}
>
  <Label
    label="Checkbox Test"
    required={false}
  />
  <InputGroup
    bsClass="input-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      inline={false}
      name="fieldName_option1"
      onChange={[Function]}
      onFocus={[Function]}
      title=""
      value="One"
    >
      Option 1
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      inline={false}
      name="fieldName_option2"
      onChange={[Function]}
      onFocus={[Function]}
      title=""
      value="Two"
    >
      Option 2
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      inline={false}
      name="fieldName_option3"
      onChange={[Function]}
      onFocus={[Function]}
      title=""
      value="Three"
    >
      Option 3
    </Checkbox>
  </InputGroup>
  <HelpBlock
    bsClass="help-block"
    style={
      Object {
        "minHeight": "6ex",
      }
    }
  >
    There was an error
    <br />
    The help Text
  </HelpBlock>
</FormGroup>
`;

exports[`The Checkbox Field renders with custom options 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="fieldName"
>
  <Label
    label="Checkbox Test"
    required={false}
  />
  <InputGroup
    bsClass="input-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      inline={false}
      name="fieldName_option1"
      onChange={[Function]}
      onFocus={[Function]}
      title=""
      value="One"
    >
      Option 1
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      inline={false}
      name="fieldName_option2"
      onChange={[Function]}
      onFocus={[Function]}
      title=""
      value="Two"
    >
      Option 2
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      inline={false}
      name="fieldName_option3"
      onChange={[Function]}
      onFocus={[Function]}
      title=""
      value="Three"
    >
      Option 3
    </Checkbox>
  </InputGroup>
  <HelpBlock
    bsClass="help-block"
    style={
      Object {
        "minHeight": "3ex",
      }
    }
  />
</FormGroup>
`;
